pxr/usdImaging/bin/usdrecord/usdrecord.cpp

```diff
@@ -0,0 +1,709 @@
+    // and test on other platforms.
+#if PXR_USE_GLFW
+    if (gpuEnabled) {
+        // UsdAppUtilsFrameRecorder will expect that an OpenGL context has
+        // been created and made current if the GPU is enabled.
+        //
+        // Frame-independent initialization.
+        // Note that the size of the widget doesn't actually affect the size of
+        // the output image. We just pass it along for cleanliness.
+
+        // Initialize the library
+        if (!glfwInit())
+            return -1;
+
+        glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE);
+
+        // Create a windowed mode window and its OpenGL context
+        GLFWwindow* window = glfwCreateWindow(args.imageWidth, args.imageWidth, "usdrecord", NULL, NULL);
+        if (!window)
+        {
+            glfwTerminate();
+            return -1;
+        }
+
+        // Make the window's context current
+        glfwMakeContextCurrent(window);
+    }
+#endif
+
+    TfToken rendererPluginId = rendererArgs_GetPluginIdFromArgument(args.rendererPlugin);
+
+    // This is the only the necessary part of RefinementComplexities python class needed for usdrecord.
+    static const std::unordered_map<std::string, float> complexities {
+            {"low", 1.0},
+            {"medium", 1.1},
+            {"high", 1.2},
+            {"veryhigh", 1.3}};
+
+    float complexity = complexities.at(args.complexity);
+
+    std::unique_ptr<UsdAppUtilsFrameRecorder> frameRecorder = std::make_unique<UsdAppUtilsFrameRecorder>(rendererPluginId, gpuEnabled);
+
+    if (!rsPrimPath.empty()) {
+        frameRecorder->SetActiveRenderSettingsPrimPath(SdfPath(rsPrimPath));
+    }
+
+    if (!args.rpPrimPath.empty()) {
+        frameRecorder->SetActiveRenderPassPrimPath(SdfPath(args.rpPrimPath));
+    }
+
+    frameRecorder->SetImageWidth(imageWidth);
+    frameRecorder->SetComplexity(complexity);
+    frameRecorder->SetCameraLightEnabled(args.cameraLightEnabled);
+    frameRecorder->SetColorCorrectionMode(colorCorrectionMode);
+    frameRecorder->SetIncludedPurposes(purposes);
+    frameRecorder->SetDomeLightVisibility(args.domeLightVisibility);
+
+    for (const auto &timeCode: frames) {
+        std::cout << "Recording time code: " << timeCode << std::endl;
+
+        std::string outputImagePath = outputImagePathPrefix;
+        if (frameNumberFormater != nullptr) {
+            // if we have a frame number formatter - then that means we successfully parsed a frame number place holder
+            // otherwise the entire original filename is in the prefix string.
+            std::stringstream ss;
+            ss << frameNumberFormater;
+            ss << timeCode << "." << outputImagePathSuffix;
+            outputImagePath += ss.str();
+        }
+
+        try {
+            frameRecorder->Record(usdStage, usdCamera, timeCode, outputImagePath);
+        } catch (TfBaseException e) {
```

Comment:
Should this be a reference?
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#e15-throw-by-value-catch-exceptions-from-a-hierarchy-by-reference

Reply:
fixed - thanks